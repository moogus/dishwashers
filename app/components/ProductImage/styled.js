import styled from 'styled-components';

export const FluidImage = styled.img`
  width: auto;
  height: 500px;
  margin-left: auto;
  margin-right: auto;
  @media (max-width: 768px) {
    margin-top:  -50px;
  }
`;

export const FluidImageContainer = styled.div`
width: 100%;
height: 500px;
  @media (max-width: 768px) {
    margin-top: 20px;
    height: 375px;
    overflow: hidden;
  }
`;

export const ActiveDotButton = styled.button`
  color: #000;
  font-size: 30px;
  border: none;
  background-color: #fff;
  padding: 0 5px 0 5px;
  margin: 0;
  &:focus {
      outline: 0 !important;
    }
`;

export const NotActiveDotButton = styled.button`
  color: #bbb;
  font-size: 30px;
  border: none;
  background-color: #fff;
  padding: 0 5px 0 5px;
  margin: 0;
  &:focus {
      outline: 0 !important;
    }
`;
