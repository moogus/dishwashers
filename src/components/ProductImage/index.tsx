import React, { useState, useEffect } from "react";

import useGetProduct from "../../hooks/product/get-product";

import {
  FluidImage,
  FluidImageContainer,
  ActiveDotButton,
  NotActiveDotButton,
} from "./styled";

export const ProductImage = () => {
  const { media } = useGetProduct();
  const [imageUrls, setImageUrls] = useState([]);
  const [selectedImage, setSelectedImage] = useState("");

  const { altText = "", urls = [] } = media;

  useEffect(() => {
    setImageUrls(urls.length < 3 ? urls.slice() : urls.slice(0, 3));
    setSelectedImage(urls[0]);
    // eslint-disable-next-line
  }, []);

  return (
    <>
      <div className="row">
        <FluidImageContainer className="col">
          <FluidImage
            data-testid="selected-image"
            src={selectedImage}
            alt={altText}
            className="d-flex justify-content-center"
          />
        </FluidImageContainer>
      </div>
      <div className="row">
        <div className="col">
          <div
            className="d-flex justify-content-center"
            data-testid="image-select-buttons"
          >
            {imageUrls.map((imageUrl) =>
              imageUrl === selectedImage ? (
                <ActiveDotButton key={imageUrl}>&#8226;</ActiveDotButton>
              ) : (
                <NotActiveDotButton
                  data-testid="activate-image"
                  key={imageUrl}
                  onClick={() => setSelectedImage(imageUrl)}
                >
                  &#8226;
                </NotActiveDotButton>
              )
            )}
          </div>
        </div>
      </div>
    </>
  );
};
