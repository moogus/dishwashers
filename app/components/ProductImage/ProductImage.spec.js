import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import ProductImage from 'components/ProductImage';
import useGetProduct from 'hooks/product/get-product';

jest.mock('hooks/product/get-product');

describe('<ProductImage />', () => {
  beforeEach(() => jest.clearAllMocks());
  const renderContainer = () => render(<ProductImage />);

  describe('when product has different amounts of images', () => {
    test('product has 1 image renders one button', () => {
      useGetProduct.mockReturnValue({
        media: {
          altText: 'what the dickens',
          urls: ['1'],
        },
      });
      const { getByTestId } = renderContainer();
      expect(getByTestId('image-select-buttons')).toMatchSnapshot();
    });

    test('product has 2 images renders 2 buttons', () => {
      useGetProduct.mockReturnValue({
        media: {
          altText: 'what the dickens',
          urls: ['1', '2'],
        },
      });
      const { getByTestId } = renderContainer();
      expect(getByTestId('image-select-buttons')).toMatchSnapshot();
    });

    test('product has 3 images renders 3 buttons', () => {
      useGetProduct.mockReturnValue({
        media: {
          altText: 'what the dickens',
          urls: ['1', '2', '3'],
        },
      });
      const { getByTestId } = renderContainer();
      expect(getByTestId('image-select-buttons')).toMatchSnapshot();
    });

    test('product has 4 images renders 3 buttons', () => {
      useGetProduct.mockReturnValue({
        media: {
          altText: 'what the dickens',
          urls: ['1', '2', '3', '4'],
        },
      });
      const { getByTestId } = renderContainer();
      expect(getByTestId('image-select-buttons')).toMatchSnapshot();
    });
  });

  describe('when <ProductImage /> has more than one image buttons can change the image', () => {
    test('cliecking image select buttons changes the image', async () => {
      useGetProduct.mockReturnValue({
        media: {
          altText: 'what the dickens',
          urls: ['first image', 'second image'],
        },
      });

      const { getByTestId, getAllByTestId } = renderContainer();
      // first image is loaded by default
      expect(getByTestId('selected-image')).toMatchSnapshot();

      const activateImageButtons = await getAllByTestId('activate-image');
      await userEvent.click(activateImageButtons[0]);
      // image is changed from clicking the button
      expect(getByTestId('selected-image')).toMatchSnapshot();
    });
  });
});
