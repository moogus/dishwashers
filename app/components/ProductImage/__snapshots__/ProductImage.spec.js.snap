// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductImage /> when <ProductImage /> has more than one image buttons can change the image cliecking image select buttons changes the image 1`] = `
<img
  alt="what the dickens"
  class="styled__FluidImage-lcexfs-0 gFFyIK d-flex justify-content-center"
  data-testid="selected-image"
  src="first image"
/>
`;

exports[`<ProductImage /> when <ProductImage /> has more than one image buttons can change the image cliecking image select buttons changes the image 2`] = `
<img
  alt="what the dickens"
  class="styled__FluidImage-lcexfs-0 gFFyIK d-flex justify-content-center"
  data-testid="selected-image"
  src="second image"
/>
`;

exports[`<ProductImage /> when product has different amounts of images product has 1 image renders one button 1`] = `
<div
  class="d-flex justify-content-center"
  data-testid="image-select-buttons"
>
  <button
    class="styled__ActiveDotButton-lcexfs-2 dwkzOS"
  >
    •
  </button>
</div>
`;

exports[`<ProductImage /> when product has different amounts of images product has 2 images renders 2 buttons 1`] = `
<div
  class="d-flex justify-content-center"
  data-testid="image-select-buttons"
>
  <button
    class="styled__ActiveDotButton-lcexfs-2 dwkzOS"
  >
    •
  </button>
  <button
    class="styled__NotActiveDotButton-lcexfs-3 jYcHDM"
    data-testid="activate-image"
  >
    •
  </button>
</div>
`;

exports[`<ProductImage /> when product has different amounts of images product has 3 images renders 3 buttons 1`] = `
<div
  class="d-flex justify-content-center"
  data-testid="image-select-buttons"
>
  <button
    class="styled__ActiveDotButton-lcexfs-2 dwkzOS"
  >
    •
  </button>
  <button
    class="styled__NotActiveDotButton-lcexfs-3 jYcHDM"
    data-testid="activate-image"
  >
    •
  </button>
  <button
    class="styled__NotActiveDotButton-lcexfs-3 jYcHDM"
    data-testid="activate-image"
  >
    •
  </button>
</div>
`;

exports[`<ProductImage /> when product has different amounts of images product has 4 images renders 3 buttons 1`] = `
<div
  class="d-flex justify-content-center"
  data-testid="image-select-buttons"
>
  <button
    class="styled__ActiveDotButton-lcexfs-2 dwkzOS"
  >
    •
  </button>
  <button
    class="styled__NotActiveDotButton-lcexfs-3 jYcHDM"
    data-testid="activate-image"
  >
    •
  </button>
  <button
    class="styled__NotActiveDotButton-lcexfs-3 jYcHDM"
    data-testid="activate-image"
  >
    •
  </button>
</div>
`;
